### PR TITLE
Convert everything to snapkit

### PR DESCRIPTION
**Pull request by Omarrasheed**
# WILL NOT MERGE UNTIL ALL ARE VIEWS AND VC'S ARE DONE

**Issue: #130**

This PR has become waaaaay bigger than I had expected it to be, especially because of animations and having to convert the three different kinds of layouts into Snapkit. Because of this, I'm hoping to merge this into master and continue working on this branch but have each subsequent PR be a smaller one that tackles one view or one file at a time. This will hopefully make it easier to review and easier for me to keep track of any bugs I potentially cause

So far, here is what is done so far:
* Converted *most* files to use Snapkit for layout
* **(BIG CHANGE)** Restructured the way we draw the route diagram. With the new arch, it should have better handling for live tracking updates, size changes, and pretty much any layout issues we had previously.
* Turned some unnecessary public functions/variables into private ones

Progress on transitioning to Snapkit completely (will update in issue):
### Cells:
- [x] BusStopTableViewCell
- [x] GeneralTableViewCell
- [x] LargeDetailTableViewCell
- [x] RouteTableViewCell
- [x] ServiceAlertTableViewCell
- [x] PlaceTableViewCell
- [x] SmallDetailTableViewCell

### Controllers:
- [x] AllStopsTableViewController
- [x] CustomNavigationController
- [x] FavoritesTableViewController
- [x] HomeMapViewController
- [x] HomeOptionsCardViewController
- [x] InformationViewController
- [x] OnboardingViewController
- [x] RouteDetailViewController 
- [x] RouteDetail+ContentViewController
- [x] RouteDetail+DrawerViewController
- [x] RouteDetailDrawerViewController+Extensions
- [x] RouteOptionsViewController
- [x] RouteOptionsViewController+Extensions
- [x] SearchResultsTableViewController
- [x] ServiceAlertsViewController

### Views:
- [x] BusIcon
- [x] BusLocationView (PR: #264)
- [x] Circle
- [x] DatePickerView
- [x] DetailIconView
- [x] HeaderView
- [x] LiveIndicator (PR: #263)
- [x] LoadingIndicator
- [x] PhraseLabelFooterView
- [x] RoundedShadowedView
- [x] RouteDiagram
- [x] RouteDiagramSegment
- [x] RouteLine
- [x] RouteSelectionView
- [x] SearchBarView
- [x] SummaryView
- [x] WalkWithDistanceIcon
- [x] WhatsNewHeaderVie
- [x] InformationTableHeaderView

### CODE REVIEW

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/LargeDetailTableViewCell.swift]: unwrap `attributedText`

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/LargeDetailTableViewCell.swift]: nit: delete newline

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/LargeDetailTableViewCell.swift]: can we store `direction.stops.count + 1` in some variable since we use it twice?

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/LargeDetailTableViewCell.swift]: `"\(detailLabelText) • \(timeString)"`

File: TCAT/Cells/RouteTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/RouteTableViewCell.swift:21]: can we make private

File: TCAT/Cells/RouteTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/RouteTableViewCell.swift]: can we change these to `let` where possible?

File: TCAT/Cells/RouteTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/RouteTableViewCell.swift]: why do we need `(UIViewController & RouteTableViewCellDelegate)?` instead of just `RouteTableViewCellDelegate?`

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/AllStopsTableViewController.swift]: nit: I prefer `private var allStops: [Place] = []`

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/AllStopsTableViewController.swift]: line 24-25: same nit as line 19

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/AllStopsTableViewController.swift]: delete?

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/AllStopsTableViewController.swift:143]: can we make a task for this or post in #transit-backend about this

File: TCAT/Controllers/FavoritesTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/FavoritesTableViewController.swift:123]: can we change this to a `switch` statement? prefer `switch` instead of `if/else` for enums

File: TCAT/Controllers/FavoritesTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/FavoritesTableViewController.swift]: we can add a `return` after line 125 and delete this `else` so that its not indented

File: TCAT/Controllers/HomeOptionsCardViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/HomeOptionsCardViewController.swift]: delete?

File: TCAT/Controllers/RouteDetail+ContentViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: can we move this into an `if let`

File: TCAT/Controllers/RouteDetail+DrawerViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+DrawerViewController.swift:43]: make private?

File: TCAT/Controllers/RouteDetail+DrawerViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+DrawerViewController.swift]: can we use `enumerated` & `forEach` here instead?

File: TCAT/Controllers/RouteDetail+DrawerViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+DrawerViewController.swift]: can combine this with line 58:
`if let cell = tableView.cellForRow(at: indexPath) as? LargeDetailTableViewCell, cell.isExpanded {...}`

File: TCAT/Views/BusLocationView.swift
Comment (kevinchan159) [inline TCAT/Views/BusLocationView.swift]: do we need this?

File: TCAT/Views/DetailIconView.swift
Comment (kevinchan159) [inline TCAT/Views/DetailIconView.swift:25]: can we change to `private let`?

File: TCAT/Views/RouteDiagramSegment.swift
Comment (kevinchan159) [inline TCAT/Views/RouteDiagramSegment.swift:17]: change to `private`?

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
love all the work so far!

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (young-k) [inline TCAT/Cells/LargeDetailTableViewCell.swift:55]: nit: you have `lineBreakMode` here but for `titleLabel` you have it after `textColor`

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (young-k) [inline TCAT/Controllers/AllStopsTableViewController.swift]: this is unnecessarily nested due to the prior `return`

File: TCAT/Controllers/FavoritesTableViewController.swift
Comment (young-k) [inline TCAT/Controllers/FavoritesTableViewController.swift]: delete

File: TCAT/Controllers/RouteDetail+ContentViewController.swift
Comment (young-k) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: should be a `TODO` or an issue

File: TCAT/Model/Section.swift
Comment (young-k) [inline TCAT/Model/Section.swift]: pretty sure it's more idiomatic to `if let` this rather than have a `switch` with one used case

File: TCAT/Model/Section.swift
Comment (young-k) [inline TCAT/Model/Section.swift:68]: probably delete? or you can add another newline below, @kevinchan159 is there a preference?

File: TCAT/Model/Section.swift
Comment (young-k) [inline TCAT/Model/Section.swift]: delete

File: TCAT/Views/RouteDiagramSegment.swift
Comment (young-k) [inline TCAT/Views/RouteDiagramSegment.swift]: prefer `if` instead of a single handled case in a `switch`

File: TCAT/Views/DatePickerView.swift
Comment (young-k) [inline TCAT/Views/DatePickerView.swift]: should this be a `weak var delegate`? cc @kevinchan159

**Comment by young-k** [comment]
Also remember to rebase –– @kevinchan159 we're missing branch protection on this repo!

File: TCAT/Model/Section.swift
Comment (kevinchan159) [reply TCAT/Model/Section.swift:68]: i prefer having the newline so just add a newline at the bottom

File: TCAT/Cells/RouteTableViewCell.swift
Comment (Omarrasheed) [reply TCAT/Cells/RouteTableViewCell.swift:21]: Don't we usually set the delegate from the caller? How would they access/set?

Comment (Omarrasheed) [reply TCAT/Cells/RouteTableViewCell.swift:21]: Oh wait wtf I j saw that we pass in the delegate in the configure. Should we be doing it this way, where we pass in a weak delegate to the cell/view, or should we be setting cell.delegate = self in the parent view?

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (Omarrasheed) [reply TCAT/Controllers/AllStopsTableViewController.swift:143]: https://github.com/cuappdev/ithaca-transit-backend/issues/242

File: TCAT/Controllers/FavoritesTableViewController.swift
Comment (Omarrasheed) [reply TCAT/Controllers/FavoritesTableViewController.swift:123]: Even for a single case?

UPDATE: Fixed in next comment

File: TCAT/Controllers/HomeOptionsCardViewController.swift
Comment (Omarrasheed) [reply TCAT/Controllers/HomeOptionsCardViewController.swift]: Good catch. Didn't mean to comment this actually. Ima double check the whole pr to make sure stuff like this didn't happen elsewhere.

File: TCAT/Cells/RouteTableViewCell.swift
Comment (kevinchan159) [reply TCAT/Cells/RouteTableViewCell.swift:21]: i prefer doing it within the configure and making it private b/c no other classes should be accessing the delegate

File: TCAT/Controllers/RouteDetail+DrawerViewController.swift
Comment (Omarrasheed) [reply TCAT/Controllers/RouteDetail+DrawerViewController.swift:43]: I'm going to tackle the private vars in a different PR hopefully.

**Review by kevinchan159 — APPROVED** [review]

**Review by young-k — APPROVED** [review]
Approving to unblock, i trust kevin's judgment here –– it looks like Drew set up squash & merge, so you might not need to squash?

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
oh wait didnt see your comment about the PR not being finished. i will re-review once its done

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/LargeDetailTableViewCell.swift:18]: `private`?

File: TCAT/Cells/RouteTableViewCell.swift
Comment (kevinchan159) [inline TCAT/Cells/RouteTableViewCell.swift]: is there a task made for this?

File: TCAT/Controllers/AllStopsTableViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/AllStopsTableViewController.swift]: you could add a `return` after line 152 and then remove this `else` block so that the code isn't indented

File: TCAT/Controllers/OnboardingViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/OnboardingViewController.swift]: really minor: you could do `(1...4).forEach()`

File: TCAT/Controllers/RouteDetail+ContentViewController.swift
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: create task?

File: TCAT/Controllers/RouteOptionsViewController+Extensions.swift
Comment (kevinchan159) [inline TCAT/Controllers/RouteOptionsViewController+Extensions.swift]: `return section == 1 ? Constants.TableHeaders.walking : nil`

**Review by kevinchan159 — APPROVED** [review]
approving to unblock. just check my comments before merging

File: TCAT/Cells/LargeDetailTableViewCell.swift
Comment (Omarrasheed) [reply TCAT/Cells/LargeDetailTableViewCell.swift:18]: Private vars are coming in a following diff :) stay tuned

File: TCAT/Cells/RouteTableViewCell.swift
Comment (Omarrasheed) [reply TCAT/Cells/RouteTableViewCell.swift]: #266

**Pull request merged**
